core/src/saros/net/xmpp/contact/XMPPContact.java

```diff
@@ -0,0 +1,297 @@
+import org.jivesoftware.smack.RosterEntry;
+import org.jivesoftware.smack.packet.RosterPacket;
+import org.jivesoftware.smack.packet.RosterPacket.ItemType;
+import saros.net.xmpp.JID;
+import saros.net.xmpp.XMPPConnectionService;
+import saros.net.xmpp.contact.ContactStatus.Type;
+
+/**
+ * An XMPPContact represents a contact on the XMPP roster. It contains the current presence state
+ * and extended informations like Saros Support. Objects of this class are thread-safe for all
+ * public methods and informations are updated by {@link XMPPConnectionService}.
+ *
+ * <p>All package private methods should be called single-threaded or synchronized to guarantee
+ * thread-safety.
+ *
+ * <p>This class is likely to implement a generic Contact Interface in the future.
+ */
+public class XMPPContact {
+
+  /** Bundles information about a resource of a client. */
+  private static final class Resource {
+    static final Comparator<Resource> BEST_RESOURCE_FIRST =
+        Comparator.comparing((Resource r) -> !r.sarosSupport)
+            .thenComparing(r -> r.status)
+            .thenComparing(r -> r.fqJid.getRAW());
+
+    final JID fqJid;
+    final ContactStatus status;
+    final boolean sarosSupport;
+
+    Resource(JID fqJid, ContactStatus status, boolean sarosSupport) {
+      this.fqJid = Objects.requireNonNull(fqJid);
+      if (fqJid.isBareJID()) throw new IllegalArgumentException();
+      this.status = Objects.requireNonNull(status);
+      this.sarosSupport = sarosSupport;
+    }
+
+    @Override
+    public String toString() {
+      return "Resource [jid="
+          + fqJid
+          + ", status="
+          + status
+          + ", sarosSupport="
+          + sarosSupport
+          + "]";
+    }
+  }
+
+  /**
+   * Create a new XMPPContact from RosterEntry.
+   *
+   * @param entry RosterEntry
+   * @return new XMPPContact
+   */
+  static XMPPContact from(RosterEntry entry) {
+    ContactStatus baseStatus;
+    if (entry.getStatus() == RosterPacket.ItemStatus.SUBSCRIPTION_PENDING) {
+      baseStatus = ContactStatus.TYPE_SUBSCRIPTION_PENDING;
+    } else if (entry.getType() == ItemType.none || entry.getType() == ItemType.from) {
+      /* see http://xmpp.org/rfcs/rfc3921.html chapter 8.2.1, 8.3.1 and 8.6 */
+      baseStatus = ContactStatus.TYPE_SUBSCRIPTION_CANCELED;
+    } else {
+      baseStatus = ContactStatus.TYPE_OFFLINE;
+    }
+
+    return new XMPPContact(new JID(entry.getUser()), baseStatus, entry.getName());
+  }
+
+  /** JID of this contact without resource information. */
+  private final JID baseJid;
+  /** Basic Status of this Contact, either offline, removed or subscription related */
+  private volatile ContactStatus baseStatus;
+  /** Roster provided nickname if available */
+  private volatile Optional<String> nickname;
+  /**
+   * Self sorted set providing the best resource as first item. As there is no atomic replace
+   * method, this set can be empty for short time frames and methods must handle this!
```

Comment:
The methods in this class or the caller of the public methods provided by this class ?